frontend/file/src/main/java/com/virtuslab/gitmachete/frontend/file/highlighting/MacheteAnnotator.java

```diff
@@ -80,6 +87,34 @@ private void processMacheteGeneratedEntry(MacheteGeneratedEntry macheteEntry, An
                   .format(processedBranchName))
           .range(branch).create();
     }
+
+    // update the state of the .git/machete VirtualFile so that new entry is available in the VirtualFile
+    ModalityUiUtil.invokeLaterIfNeeded(NON_MODAL, () -> saveDocumentBeforeCheck(file));
+    // check for duplicate entries in the machete file
+    try {
+      val branchNamesFromFile = VfsUtil.loadText(file.getVirtualFile());
+      if (isBranchNameRepeated(branchNamesFromFile, processedBranchName)) {
+        holder.newAnnotation(HighlightSeverity.ERROR,
+            getNonHtmlString("string.GitMachete.MacheteAnnotator.branch-entry-already-defined")
+                .format(processedBranchName))
+            .range(branch).create();
+      }
+    } catch (PluginException | IllegalStateException ignored) { // ignore dubious IDE checks against annotation range
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @UIEffect
+  private void saveDocumentBeforeCheck(PsiFile file) {
+    val fileDocManager = FileDocumentManager.getInstance();
+    fileDocManager.saveDocument(Objects.requireNonNull(fileDocManager.getDocument(file.getVirtualFile())));
+  }
+
+  private boolean isBranchNameRepeated(String fileContent, String branchName) {
+    java.util.List<String> lines = Arrays.stream(fileContent.split(System.lineSeparator())).map(String::trim)
+        .collect(Collectors.toList());
```

Comment:
Awww I see where the problem lies 🤦🏻 (at least partially)

This algorithm for detecting duplicates does NOT take into account custom annotations:
```
foo
  bar PR #123
  bar PR #456
```
Then `isBranchNameRepeated(..., "bar")` will return 0️⃣  (as there are no lines which contain exactly `bar` after trimming).

I'd suggest the following:
1. if possible, rely on **parsed data** delivered by IntelliJ APIs, rather than do rudimentary parsing directly in this method

and/or

2. given that a non-trivial bug has been detected on this PR, pls add a regression test for the logic of detecting the duplicates (unless that requires heavily mocking IntelliJ APIs)